docs/writing-tests/testharness-tutorial.md

```diff
@@ -0,0 +1,390 @@
+
+## Refining the subtest
+
+<!--
+Goals of this section:
+
+- explain the motivation for "clean up" logic and demonstrate its usage
+- motivate explicit test naming
+-->
+
+We'd like to test a little more about `fetch` and cookies, but before we do,
+there are some improvements we can make to what we've written so far.
+
+For instance, we should remove the cookie after the subtest is complete. This
+ensures a consistent state for any additional subtests we may add and also for
+any tests that follow. We'll use the `add_cleanup` method to ensure that the
+cookie is deleted even if the test fails.
+
+```diff
+-promise_test(function() {
++promise_test(function(t) {
++  t.add_cleanup(function() {
++    document.cookie = 'test1=;expires=Thu, 01 Jan 1970 00:00:01 GMT;';
++  });
++
+   return fetch('thing.asis')
+     .then(function() {
+         assert_equals(document.cookie, 'test1=t1');
+       });
+ });
+```
+
+Although we'd prefer it if there were no other cookies defined during our test,
+we shouldn't take that for granted. As written, the test will fail if the
+`document.cookie` includes additional cookies. We'll use slightly more
+complicated logic to test for the presence of the expected cookie.
+
+
+```diff
+ promise_test(function(t) {
+   t.add_cleanup(function() {
+     document.cookie = 'test1=;expires=Thu, 01 Jan 1970 00:00:01 GMT;';
+   });
+
+   return fetch('thing.asis')
+     .then(function() {
+-        assert_equals(document.cookie, 'test1=t1');
++        assert_true(/(^|; )test1=t1($|;)/.test(document.cookie);
+       });
+ });
+```
+
+In the screen shot above, the subtest's result was reported using the
+document's title, "fetch: setting cookies". Since we expect to add another
+subtest, we should give this one a more specific name:
+
+```diff
+ promise_test(function(t) {
+   t.add_cleanup(function() {
+     document.cookie = 'test1=;expires=Thu, 01 Jan 1970 00:00:01 GMT;';
+   });
+
+   return fetch('thing.asis')
+     .then(function() {
+         assert_true(/(^|; )test1=t1($|;)/.test(document.cookie));
+       });
+-});
++}, 'cookie set for successful request');
+```
+
+## Writing a second subtest
+
+<!--
+Goals of this section:
+
+- introduce the concept of cross-domain testing and the associated tooling
+- demonstrate how to verify promise rejection
+- demonstrate additional assertion functions
+-->
+
+There are many things we might want to verify about how `fetch` sets cookies.
+For instance, it should *not* set a cookie if the request fails due to
+cross-origin security restrictions. Let's write a subtest which verifies that.
+
+We'll add another `<script>` tag for a JavaScript support file:
+
+```diff
+ <!DOCTYPE html>
+ <meta charset="utf-8">
+ <title>fetch: setting cookies</title>
+ <script src="/resources/testharness.js"></script>
+ <script src="/resources/testharnessreport.js"></script>
++<script src="/common/get-host-info.sub.js"></script>
```

Comment:
> maybe explain what extra functionality we're using from this file?

It seems to me that the tutorial already has this information:

> `get-host-info.sub.js` is a general-purpose script provided by WPT. It's designed to help with testing cross-domain functionality.

And later (in context of its usage):

> The name of the host depends on how the WPT server has been configured, so we rely on the helper to provide an appropriate value.

Do you think something is missing? Or maybe that we should just reorganize 

> Is there a place that explains the functionality that's available, and which files you can find it in?

There is not, and that is an excellent point. Writing documentation for it will enhance this tutorial, but the problem is much larger, so I'd like to address it in a separate effort. I filed gh-17913 to report the problem and track its resolution.

Reply:
Sounds good. Yes, I think the existing explanations are adequate for this particular case (I don't think I read this bit thoroughly enough)